packages/react/checkbox/src/Checkbox.tsx

```diff
@@ -54,72 +55,60 @@ const Checkbox = React.forwardRef((props, forwardedRef) => {
     onCheckedChange,
     ...checkboxProps
   } = props;
-  const inputRef = React.useRef<HTMLInputElement>(null);
   const [button, setButton] = React.useState<HTMLButtonElement | null>(null);
   const composedRefs = useComposedRefs(forwardedRef, (node) => setButton(node));
   const buttonSize = useSize(button);
   const labelId = useLabelContext(button);
   const labelledBy = ariaLabelledby || labelId;
+  const hasConsumerStoppedPropagationRef = React.useRef(false);
   const [checked = false, setChecked] = useControllableState({
     prop: checkedProp,
     defaultProp: defaultChecked,
-  });
-
-  React.useEffect(() => {
-    const isIndeterminate = checked === 'indeterminate';
-    inputRef.current && (inputRef.current.indeterminate = isIndeterminate);
+    onChange: onCheckedChange,
   });
 
   return (
-    /**
-     * The `input` is hidden from non-SR and SR users as it only exists to
-     * ensure form events fire when the value changes and that the value
-     * updates when clicking an associated label.
-     */
-    <>
-      <input
-        ref={inputRef}
-        type="checkbox"
+    <CheckboxProvider state={checked} disabled={disabled}>
+      <Primitive
+        type="button"
+        role="checkbox"
+        aria-checked={checked === 'indeterminate' ? 'mixed' : checked}
+        aria-labelledby={labelledBy}
+        aria-required={required}
+        data-state={getState(checked)}
+        data-disabled={disabled ? '' : undefined}
+        disabled={disabled}
+        value={value}
+        {...checkboxProps}
+        as={as}
+        ref={composedRefs}
+        onClick={composeEventHandlers(props.onClick, (event) => {
+          setChecked((prevChecked) => (prevChecked === 'indeterminate' ? true : !prevChecked));
+          hasConsumerStoppedPropagationRef.current = event.isPropagationStopped();
+          // stop propagation from the button so that we only propagate one click event (from the input)
+          event.stopPropagation();
```

Comment:
You may need to run me through this routine here, because I don't think I'm following, it seems we always stop propagation…